README.md
[![GitHub release](https://img.shields.io/github/release/QzSG/bluemix-ghost-cloudinary.svg?style=flat-square)](https://github.com/QzSG/bluemix-ghost-cloudinary/releases)
[![Ghost Version](https://img.shields.io/badge/ghost-0.11.10-brightgreen.svg?maxAge=2592000&style=flat-square)](https://github.com/TryGhost/Ghost)
[![Node Version](https://img.shields.io/badge/node-^4.5-brightgreen.svg?maxAge=2592000&style=flat-square)](https://nodejs.org/en/)
[![Travis for Ghost](https://img.shields.io/travis/TryGhost/Ghost/master.svg?maxAge=2592000&style=flat-square)](https://travis-ci.org/TryGhost/Ghost)
[![David dependencies checker](https://img.shields.io/david/QzSG/bluemix-ghost-cloudinary.svg?maxAge=2592000&style=flat-square)](https://david-dm.org/QzSG/bluemix-ghost-cloudinary)
[![MIT license](https://img.shields.io/github/license/QzSG/bluemix-ghost-cloudinary.svg?maxAge=2592000&style=flat-square)](https://github.com/QzSG/bluemix-ghost-cloudinary/blob/master/LICENSE)

<img src="https://cloud.githubusercontent.com/assets/120485/6622822/c4c639fe-c8e7-11e4-9e64-5bec06c8b4c3.png" alt="Ghost" />

![Ghost Screenshot](https://cloud.githubusercontent.com/assets/120485/6626466/6dae46b2-c8ff-11e4-8c7c-8dd63b215f7b.jpg)

![Ghost is a simple, powerful publishing platform that allows you to share your stories with the world.](https://cloud.githubusercontent.com/assets/120485/6626501/b2bb072c-c8ff-11e4-8e1a-2e78e68fd5c3.png)

The project is maintained by a non-profit organisation called the **Ghost Foundation**, along with an amazing group of independent [contributors](https://github.com/TryGhost/Ghost/contributors). We're trying to make publishing software that changes the shape of online journalism.

- [Visit Ghost.org for more details](https://ghost.org)

# Quick Start Install

Make sure you've installed Node.js - I recommend the latest **Node v4 LTS** release. For other supported versions [click here](http://support.ghost.org/supported-node-versions/). 

Install Node.js (if you haven't done so) 

Ghost currently supports Node versions 4.5+ and 6.9+ only.

Clone :ghost:

```bash
git clone git://github.com/qzsg/bluemix-ghost-cloudinary.git ghost-blog
cd ghost-blog
```

Install Ghost and dependencies

```bash
npm install
```

Start your engines.

```bash
npm start

## running production? Add --production
```

Congrats! You made it. Visit your local Ghost blog at `http://localhost:2368/` :tada:

# Developer Install (from git)

Install Node.js. (See [Supported Node.js versions](http://support.ghost.org/supported-node-versions/))


# Deploying Ghost on BlueMix

![Ghost + BlueMix](http://i.imgur.com/vNrQ3PD.jpg?1)

<a href="https://bluemix.net/deploy?repository=https://github.com/QzSG/bluemix-ghost-cloudinary"><img src="https://bluemix.net/deploy/button.svg" alt="Deploy to Bluemix" width="300px"/></a>

### Storage: Using [Cloudinary](http://cloudinary.com/) to store files

1. After you have created an account at Cloudinary
2. Uncomment the Cloudinary configuration inside config.js
3. Create an environment variable containing your cloudinary credentials which can be found at [Cloudinary Console](https://cloudinary.com/console), in the format of: `cf set-env ghost CLOUDINARY '{"cloud_name":"...", "api_key":"...", "api_secret":"..."}'`
4. Push update and restart ghost `cf push`



# Helpful references
- [TryGhost/Ghost](https://github.com/TryGhost/Ghost)
- [sethbrasile/ghost-cloudinary-store](https://github.com/sethbrasile/ghost-cloudinary-store)
- [ibmjstart/bluemix-ghost-js](https://github.com/ibmjstart/bluemix-ghost-js)

# Copyright & License

Copyright (c) QzSG - Released under the [MIT license](LICENSE).


CHANGELOG.md
# Change Log

## [1.3.0](https://github.com/QzSG/bluemix-ghost-cloudinary/tree/1.3.0) (2017-06-17)
[Full Changelog](https://github.com/QzSG/bluemix-ghost-cloudinary/compare/1.2.8...1.3.0)

## [1.2.8](https://github.com/QzSG/bluemix-ghost-cloudinary/tree/1.2.8) (2017-05-04)
[Full Changelog](https://github.com/QzSG/bluemix-ghost-cloudinary/compare/1.2.7...1.2.8)

## [1.2.7](https://github.com/QzSG/bluemix-ghost-cloudinary/tree/1.2.7) (2017-03-22)
[Full Changelog](https://github.com/QzSG/bluemix-ghost-cloudinary/compare/1.2.3...1.2.7)

## [1.2.3](https://github.com/QzSG/bluemix-ghost-cloudinary/tree/1.2.3) (2016-11-06)
[Full Changelog](https://github.com/QzSG/bluemix-ghost-cloudinary/compare/1.2.2...1.2.3)

## [1.2.2](https://github.com/QzSG/bluemix-ghost-cloudinary/tree/1.2.2) (2016-10-15)
[Full Changelog](https://github.com/QzSG/bluemix-ghost-cloudinary/compare/1.2.1...1.2.2)

**Implemented enhancements:**

- Add Contributing.md [\#6](https://github.com/QzSG/bluemix-ghost-cloudinary/issues/6)

## [1.2.1](https://github.com/QzSG/bluemix-ghost-cloudinary/tree/1.2.1) (2016-10-05)
[Full Changelog](https://github.com/QzSG/bluemix-ghost-cloudinary/compare/1.1.1...1.2.1)

**Merged pull requests:**

- Update Ghost to version 0.11.1 [\#8](https://github.com/QzSG/bluemix-ghost-cloudinary/pull/8) ([QzSG](https://github.com/QzSG))
- Add Contributing.md [\#7](https://github.com/QzSG/bluemix-ghost-cloudinary/pull/7) ([QzSG](https://github.com/QzSG))
- Updated changelog [\#5](https://github.com/QzSG/bluemix-ghost-cloudinary/pull/5) ([QzSG](https://github.com/QzSG))

## [1.1.1](https://github.com/QzSG/bluemix-ghost-cloudinary/tree/1.1.1) (2016-09-18)
[Full Changelog](https://github.com/QzSG/bluemix-ghost-cloudinary/compare/1.1.0...1.1.1)

**Closed issues:**

- Bump minimum NODE & NPM versions [\#3](https://github.com/QzSG/bluemix-ghost-cloudinary/issues/3)
- IBM DevOps Pipeline running tasks twice [\#1](https://github.com/QzSG/bluemix-ghost-cloudinary/issues/1)

**Merged pull requests:**

- Develop [\#4](https://github.com/QzSG/bluemix-ghost-cloudinary/pull/4) ([QzSG](https://github.com/QzSG))
- Prevent `npm install & preinstall` from running twice [\#2](https://github.com/QzSG/bluemix-ghost-cloudinary/pull/2) ([QzSG](https://github.com/QzSG))

## [1.1.0](https://github.com/QzSG/bluemix-ghost-cloudinary/tree/1.1.0) (2016-09-18)
[Full Changelog](https://github.com/QzSG/bluemix-ghost-cloudinary/compare/1.0.0...1.1.0)

## [1.0.0](https://github.com/QzSG/bluemix-ghost-cloudinary/tree/1.0.0) (2016-09-17)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*


config.js
// # Ghost Configuration
// Setup your Ghost install for various [environments](http://support.ghost.org/config/#about-environments).

// Ghost runs in `development` mode by default. Full documentation can be found at http://support.ghost.org/config/

var path = require('path'),
    config;

// my app config
var myapp = {
    //Uncomment the below if you have cloudinary account at https://cloudinary.com/ created and added credentials to node env
    //cloudinary: {},
    //Uncomment the below if you have mailgun account at https://www.mailgun.com/ created and added credentials to node env
    //mailgun: {},
    mysql: {}
};
myapp.port = (process.env.VCAP_APP_PORT || '2368');
myapp.host = (process.env.VCAP_APP_HOST || '127.0.0.1');
myapp.protocol = 'https://';

if (process.env.VCAP_APPLICATION != undefined) {
    var vcap_application = JSON.parse(process.env.VCAP_APPLICATION);
    myapp.route = vcap_application.application_uris[0];
}

if (process.env.VCAP_SERVICES != undefined) {
    var vcap_services = JSON.parse(process.env.VCAP_SERVICES);
    
    if (vcap_services.mysql != undefined) {
        myapp.mysql = vcap_services.mysql[0].credentials;
    } else {
        console.error('Cannot find mysql service');
        process.exit(1);
    }
}

//Uncomment the below if you have cloudinary account at https://cloudinary.com/ created and added credentials to node env
/*if (process.env.CLOUDINARY != undefined)
    myapp.cloudinary = JSON.parse(process.env.CLOUDINARY);

//console.log(myapp.cloudinary);
*/

//Uncomment the below if you have mailgun account at https://www.mailgun.com/ created and added credentials to node env
/*if (process.env.MAILGUN != undefined)
    myapp.mailgun = JSON.parse(process.env.MAILGUN);
*/


// end: my app config


config = {
    // ### Production
    // When running Ghost in the wild, use the production environment.
    // Configure your URL and mail settings here
    production: {
        url: myapp.protocol + myapp.route,
        // Uncomment mail block once you have gotten mailgun credentials and added env variables
        /*mail: {
            transport: 'SMTP',
            options: {
                service: 'Mailgun',
                auth: {
                    user: myapp.mailgun.user, // mailgun username
                    pass: myapp.mailgun.pass  // mailgun password
                }
            }
        },
        */
        database: {
            client: 'mysql',
            connection: {
                host: myapp.mysql.hostname,
                port: myapp.mysql.port,
                user: myapp.mysql.username,
                password: myapp.mysql.password,
                database: myapp.mysql.name,
                charset: 'utf8'
            },
            pool: {
                min: 2,
                max: 2
            }
        },
        //Uncomment the below if you have cloudinary account at https://cloudinary.com/ created and added credentials to node env
        // Note that local storage will not persist
        /*
        storage: {
            active: 'ghost-cloudinary-store',
            'ghost-cloudinary-store': {
                cloud_name: myapp.cloudinary.cloud_name,
                api_key: myapp.cloudinary.api_key,
                api_secret: myapp.cloudinary.api_secret,
                secure: true
            }
        },
        */
        server: {
            // Host to be passed to node's `net.Server#listen()`
            host: myapp.host,
            // Port to be passed to node's `net.Server#listen()`, for iisnode set this to `process.env.PORT`
            port: myapp.port
        },
        forceAdminSSL: true
    },

    // ### Development **(default)**
    development: {
        // The url to use when providing links to the site, E.g. in RSS and email.
        // Change this to your Ghost blog's published URL.
        url: 'http://localhost:2368',

        // Example refferer policy
        // Visit https://www.w3.org/TR/referrer-policy/ for instructions
        // default 'origin-when-cross-origin',
        // referrerPolicy: 'origin-when-cross-origin',

        // Example mail config
        // Visit http://support.ghost.org/mail for instructions
        // ```
        //  mail: {
        //      transport: 'SMTP',
        //      options: {
        //          service: 'Mailgun',
        //          auth: {
        //              user: '', // mailgun username
        //              pass: ''  // mailgun password
        //          }
        //      }
        //  },
        // ```

        // #### Database
        // Ghost supports sqlite3 (default), MySQL & PostgreSQL
        database: {
            client: 'sqlite3',
            connection: {
                filename: path.join(__dirname, '/content/data/ghost-dev.db')
            },
            debug: false
        },
        // #### Server
        // Can be host & port (default), or socket
        server: {
            // Host to be passed to node's `net.Server#listen()`
            host: '127.0.0.1',
            // Port to be passed to node's `net.Server#listen()`, for iisnode set this to `process.env.PORT`
            port: '2368'
        },
        // #### Paths
        // Specify where your content directory lives
        paths: {
            contentPath: path.join(__dirname, '/content/')
        }
    },

    // **Developers only need to edit below here**

    // ### Testing
    // Used when developing Ghost to run tests and check the health of Ghost
    // Uses a different port number
    testing: {
        url: 'http://127.0.0.1:2369',
        database: {
            client: 'sqlite3',
            connection: {
                filename: path.join(__dirname, '/content/data/ghost-test.db')
            },
            pool: {
                afterCreate: function (conn, done) {
                    conn.run('PRAGMA synchronous=OFF;' +
                    'PRAGMA journal_mode=MEMORY;' +
                    'PRAGMA locking_mode=EXCLUSIVE;' +
                    'BEGIN EXCLUSIVE; COMMIT;', done);
                }
            },
            useNullAsDefault: true
        },
        server: {
            host: '127.0.0.1',
            port: '2369'
        },
        logging: false
    },

    // ### Testing MySQL
    // Used by Travis - Automated testing run through GitHub
    'testing-mysql': {
        url: 'http://127.0.0.1:2369',
        database: {
            client: 'mysql',
            connection: {
                host     : '127.0.0.1',
                user     : 'root',
                password : '',
                database : 'ghost_testing',
                charset  : 'utf8'
            }
        },
        server: {
            host: '127.0.0.1',
            port: '2369'
        },
        logging: false
    },

    // ### Testing pg
    // Used by Travis - Automated testing run through GitHub
    'testing-pg': {
        url: 'http://127.0.0.1:2369',
        database: {
            client: 'pg',
            connection: {
                host     : '127.0.0.1',
                user     : 'postgres',
                password : '',
                database : 'ghost_testing',
                charset  : 'utf8'
            }
        },
        server: {
            host: '127.0.0.1',
            port: '2369'
        },
        logging: false
    }
};

module.exports = config;


content/storage/ghost-cloudinary-store/index.js
var Promise = require('bluebird');
var cloudinary = require('cloudinary');
var util = require('util');

baseStore = require('../../../core/server/storage/base');

// TODO: Add support for private_cdn
// TODO: Add support for secure_distribution
// TODO: Add support for cname
// TODO: Add support for cdn_subdomain
// http://cloudinary.com/documentation/node_additional_topics#configuration_options

function CloudinaryStore(config) {
    baseStore.call(this);
    this.config = config || {};
    cloudinary.config(config);
}


util.inherits(CloudinaryStore, baseStore);

CloudinaryStore.prototype.save = function (image) {
    var secure = this.config.secure || false;
    
    return new Promise(function (resolve) {
        cloudinary.uploader.upload(image.path, function (result) {
            resolve(secure ? result.secure_url : result.url);
        });
    });
};

CloudinaryStore.prototype.delete = function (image) {
    
    return new Promise(function (resolve) {
        cloudinary.uploader.destroy('zombie', function (result) {
            resolve(result)
        });
    });
};

CloudinaryStore.prototype.exists = function (filename) {
    return new Promise(function (resolve) {
        if (cloudinary.image(filename, {})) {
            resolve(true);
        } else {
            resolve(false);
        }
    });

}

CloudinaryStore.prototype.serve = function () {
    return function (req, res, next) {
        next();
    };
};

module.exports = CloudinaryStore;

config.example.js
// # Ghost Configuration
// Setup your Ghost install for various [environments](http://support.ghost.org/config/#about-environments).

// Ghost runs in `development` mode by default. Full documentation can be found at http://support.ghost.org/config/

var path = require('path'),
    config;

// my app config
var myapp = {
    //Uncomment the below if you have cloudinary account at https://cloudinary.com/ created and added credentials to node env
    //cloudinary: {},
    //Uncomment the below if you have mailgun account at https://www.mailgun.com/ created and added credentials to node env
    //mailgun: {},
    mysql: {}
};
myapp.port = (process.env.VCAP_APP_PORT || '2368');
myapp.host = (process.env.VCAP_APP_HOST || '127.0.0.1');
myapp.protocol = 'https://';

if (process.env.VCAP_APPLICATION != undefined) {
    var vcap_application = JSON.parse(process.env.VCAP_APPLICATION);
    myapp.route = vcap_application.application_uris[0];
}

if (process.env.VCAP_SERVICES != undefined) {
    var vcap_services = JSON.parse(process.env.VCAP_SERVICES);
    
    /* //uncomment if using experimental mysql service
    if (vcap_services.mysql != undefined) {
        myapp.mysql = vcap_services.mysql[0].credentials;
    } else {
        console.error('Cannot find mysql service');
        process.exit(1);
    }*/
    
    // Default uses ClearDB
    if (vcap_services.cleardb != undefined) {
        myapp.mysql = vcap_services.cleardb[0].credentials;
    } else {
        console.error('Cannot find cleardb service');
        process.exit(1);
    }
}

//Uncomment the below if you have cloudinary account at https://cloudinary.com/ created and added credentials to node env
/*if (process.env.CLOUDINARY != undefined)
    myapp.cloudinary = JSON.parse(process.env.CLOUDINARY);

//console.log(myapp.cloudinary);
*/

//Uncomment the below if you have mailgun account at https://www.mailgun.com/ created and added credentials to node env
/*if (process.env.MAILGUN != undefined)
    myapp.mailgun = JSON.parse(process.env.MAILGUN);
*/


// end: my app config


config = {
    // ### Production
    // When running Ghost in the wild, use the production environment.
    // Configure your URL and mail settings here
    production: {
        url: myapp.protocol + myapp.route,
        // Uncomment mail block once you have gotten mailgun credentials and added env variables
        /*mail: {
            transport: 'SMTP',
            options: {
                service: 'Mailgun',
                auth: {
                    user: myapp.mailgun.user, // mailgun username
                    pass: myapp.mailgun.pass  // mailgun password
                }
            }
        },
        */
        database: {
            client: 'mysql',
            connection: {
                host: myapp.mysql.hostname,
                port: myapp.mysql.port,
                user: myapp.mysql.username,
                password: myapp.mysql.password,
                database: myapp.mysql.name,
                charset: 'utf8'
            },
            pool: {
                min: 2,
                max: 2
            }
        },
        //Uncomment the below if you have cloudinary account at https://cloudinary.com/ created and added credentials to node env
        // Note that local storage will not persist
        /*
        storage: {
            active: 'ghost-cloudinary-store',
            'ghost-cloudinary-store': {
                cloud_name: myapp.cloudinary.cloud_name,
                api_key: myapp.cloudinary.api_key,
                api_secret: myapp.cloudinary.api_secret,
                secure: true
            }
        },
        */
        server: {
            // Host to be passed to node's `net.Server#listen()`
            host: myapp.host,
            // Port to be passed to node's `net.Server#listen()`, for iisnode set this to `process.env.PORT`
            port: myapp.port
        },
        forceAdminSSL: true
    },

    // ### Development **(default)**
    development: {
        // The url to use when providing links to the site, E.g. in RSS and email.
        // Change this to your Ghost blog's published URL.
        url: 'http://localhost:2368',

        // Example refferer policy
        // Visit https://www.w3.org/TR/referrer-policy/ for instructions
        // default 'origin-when-cross-origin',
        // referrerPolicy: 'origin-when-cross-origin',

        // Example mail config
        // Visit http://support.ghost.org/mail for instructions
        // ```
        //  mail: {
        //      transport: 'SMTP',
        //      options: {
        //          service: 'Mailgun',
        //          auth: {
        //              user: '', // mailgun username
        //              pass: ''  // mailgun password
        //          }
        //      }
        //  },
        // ```

        // #### Database
        // Ghost supports sqlite3 (default), MySQL & PostgreSQL
        database: {
            client: 'sqlite3',
            connection: {
                filename: path.join(__dirname, '/content/data/ghost-dev.db')
            },
            debug: false
        },
        // #### Server
        // Can be host & port (default), or socket
        server: {
            // Host to be passed to node's `net.Server#listen()`
            host: '127.0.0.1',
            // Port to be passed to node's `net.Server#listen()`, for iisnode set this to `process.env.PORT`
            port: '2368'
        },
        // #### Paths
        // Specify where your content directory lives
        paths: {
            contentPath: path.join(__dirname, '/content/')
        }
    },

    // **Developers only need to edit below here**

    // ### Testing
    // Used when developing Ghost to run tests and check the health of Ghost
    // Uses a different port number
    testing: {
        url: 'http://127.0.0.1:2369',
        database: {
            client: 'sqlite3',
            connection: {
                filename: path.join(__dirname, '/content/data/ghost-test.db')
            },
            pool: {
                afterCreate: function (conn, done) {
                    conn.run('PRAGMA synchronous=OFF;' +
                    'PRAGMA journal_mode=MEMORY;' +
                    'PRAGMA locking_mode=EXCLUSIVE;' +
                    'BEGIN EXCLUSIVE; COMMIT;', done);
                }
            },
            useNullAsDefault: true
        },
        server: {
            host: '127.0.0.1',
            port: '2369'
        },
        logging: false
    },

    // ### Testing MySQL
    // Used by Travis - Automated testing run through GitHub
    'testing-mysql': {
        url: 'http://127.0.0.1:2369',
        database: {
            client: 'mysql',
            connection: {
                host     : '127.0.0.1',
                user     : 'root',
                password : '',
                database : 'ghost_testing',
                charset  : 'utf8'
            }
        },
        server: {
            host: '127.0.0.1',
            port: '2369'
        },
        logging: false
    },

    // ### Testing pg
    // Used by Travis - Automated testing run through GitHub
    'testing-pg': {
        url: 'http://127.0.0.1:2369',
        database: {
            client: 'pg',
            connection: {
                host     : '127.0.0.1',
                user     : 'postgres',
                password : '',
                database : 'ghost_testing',
                charset  : 'utf8'
            }
        },
        server: {
            host: '127.0.0.1',
            port: '2369'
        },
        logging: false
    }
};

module.exports = config;
